carbonserver/fetchsinglemetric.go

```diff
@@ -96,44 +98,63 @@ func (listener *CarbonserverListener) fetchSingleMetric(metric string, pathExpre
 		zap.Int("untilTime", int(untilTime)),
 	)
 	m, err := listener.fetchFromDisk(metric, fromTime, untilTime)
-	if err != nil {
-		atomic.AddUint64(&listener.metrics.RenderErrors, 1)
-		logger.Warn("failed to fetch points", zap.Error(err))
-		return response{}, err
-	}
+	if err == nil {
+		// Should never happen, because we have a check for proper archive now
+		if m.Timeseries == nil {
+			atomic.AddUint64(&listener.metrics.RenderErrors, 1)
+			logger.Warn("metric time range not found")
+			return response{}, errors.New("time range not found")
+		}
 
-	// Should never happen, because we have a check for proper archive now
-	if m.Timeseries == nil {
-		atomic.AddUint64(&listener.metrics.RenderErrors, 1)
-		logger.Warn("metric time range not found")
-		return response{}, errors.New("time range not found")
-	}
+		values := m.Timeseries.Values()
+		from := int64(m.Timeseries.FromTime())
+		until := int64(m.Timeseries.UntilTime())
+		step := int64(m.Timeseries.Step())
+
+		resp := response{
+			Name:              metric,
+			StartTime:         from,
+			StopTime:          until,
+			StepTime:          step,
+			Values:            values,
+			PathExpression:    pathExpression,
+			ConsolidationFunc: m.Metadata.ConsolidationFunc,
+			XFilesFactor:      m.Metadata.XFilesFactor,
+			RequestStartTime:  int64(fromTime),
+			RequestStopTime:   int64(untilTime),
+		}
 
-	values := m.Timeseries.Values()
-	from := int64(m.Timeseries.FromTime())
-	until := int64(m.Timeseries.UntilTime())
-	step := int64(m.Timeseries.Step())
-
-	resp := response{
-		Name:              metric,
-		StartTime:         from,
-		StopTime:          until,
-		StepTime:          step,
-		Values:            values,
-		PathExpression:    pathExpression,
-		ConsolidationFunc: m.Metadata.ConsolidationFunc,
-		XFilesFactor:      m.Metadata.XFilesFactor,
-		RequestStartTime:  int64(fromTime),
-		RequestStopTime:   int64(untilTime),
-	}
+		resp.enrichFromCache(listener, m.CacheData)
 
-	resp.enrichFromCache(listener, m)
+		logger.Debug("fetched",
+			zap.Any("response", resp),
+		)
 
-	logger.Debug("fetched",
-		zap.Any("response", resp),
-	)
+		return resp, nil
+	} else if os.IsNotExist(err) && listener.cacheGetRecentMetrics != nil {
+		//Failed to fetch from disk, try to fetch from cache
+		resp := response{
+			Name:           metric,
+			PathExpression: pathExpression,
+		}
+		cacheData, cerr := listener.fetchFromCache(metric, fromTime, untilTime, &resp)
+		if cerr != nil {
+			atomic.AddUint64(&listener.metrics.RenderErrors, 1)
+			logger.Warn("metric not found even in Cache", zap.Error(err))
+			// Metric has no Whisper file and/or has no datapoints in Cache
+			return response{}, cerr
+		}
+		resp.enrichFromCache(listener, cacheData)
```

Comment:
is this still necessary? Since it's fetched from cache already.

Reply:
Yes. The actual data is populated in the call to enrichFromCache.